kernel/src/vm/vmar/vm_mapping.rs

```diff
@@ -314,8 +359,20 @@ impl VmMapping {
         write: bool,
     ) -> core::result::Result<(UFrame, bool), VmoCommitError> {
         let mut is_readonly = false;
-        let Some(vmo) = &self.vmo else {
-            return Ok((FrameAllocOptions::new().alloc_frame()?.into(), is_readonly));
+
+        let vmo = match &self.mapped_mem {
+            MappedMemory::Vmo(vmo) => vmo,
+            MappedMemory::Anonymous => {
+                // Anonymous mapping. Allocate a new frame.
+                return Ok((FrameAllocOptions::new().alloc_frame()?.into(), is_readonly));
+            }
+            MappedMemory::Device => {
+                // Device memory should not go through the page fault handler
+                return Err(VmoCommitError::Err(Error::with_message(
+                    Errno::EFAULT,
+                    "Device memory mapping should not go through the page fault handler",
+                )));
```

Comment:
The wording is inaccurate:
```suggestion
                // Device memory is populated when the memory mapping is created.
                return Err(VmoCommitError::Err(Error::with_message(
                    Errno::EFAULT,
                    "device memory page faults cannot be resolved",
                )));
```